fetch_and_insert_vpd_rings()

SergiiDmytruk opened the pull request:
Not sure those intermediate buffers can be avoided. When applying an overlay, all three of them are used at the same time.

I've left "proto-"`getPPeScanRings()` which I used for testing there to show how new function is to be used.

RS4 implementation is taken from Hostboot. It's about 600 lines of error-prone code and we have no test suite. Rewriting it is reasonable only after we can test the implementation by using it extensively.

MVPD might be using wrong data area to back up `mmap` operation, just don't know what else to use there.

SergiiDmytruk committed:
soc/power9/vpd.c: fix some comments

Change-Id: If68733b6d0535589f6d997df1a209f200eff689a
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
soc/power9/vpd.c: export vpd_find_kwd()

Change-Id: If418564d2498a06b3fd6fb426016d1b82809ee0e
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
soc/power9/rom_media.c: MVPD partition

Change-Id: I1312410e76376367cc61b5f04306e77df369dc26
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
soc/power9/mvpd.c: implement mvpd_extract_ring()

Change-Id: I29e9b193af5f001d5a1f87f538e421d74700c735
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
soc/power9/tor.c: implement basic tor functions

Change-Id: Ic3b6a40d0d998c925d5bddf034c9314e4bc13e1f
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
soc/power9/homer.c: proto-getPpeScanRings()

Change-Id: Ibcbf19b5621554c144962c1246c3fe518ddb4afb
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -885,6 +887,42 @@ static void istep_16_1(int this_core)
+	static uint8_t buf3[MAX_RING_BUF_SIZE];
+
+	tor_access_ring((struct tor_hdr *)(homer_bar + hw->rings.offset + 68),
+			EC_TIME, PT_CME, 0, ppe, &ppe_size, GET_PPE_LEVEL_RINGS);
```
```suggestion
	struct tor_hdr *tor;
	copy_section(&tor, &hw->rings, hw, dd, FIND);
	tor_access_ring(tor,
			EC_TIME, PT_CME, 0, ppe, &ppe_size, GET_PPE_LEVEL_RINGS);
```

Just FYI how to use `copy_section()`. It will choose TOR for proper `dd`, which should be passed as an argument to `getPpeScanRings()`.

krystian-hebel commented on src/soc/ibm/power9/mvpd.c:
```diff
@@ -0,0 +1,158 @@
+#include "rs4.h"
+
+#define MVPD_TOC_ENTRIES 32
+#define MVPD_TOC_SIZE    MVPD_TOC_ENTRIES*sizeof(struct mvpd_toc_entry)
```
```suggestion
#define MVPD_TOC_SIZE    (MVPD_TOC_ENTRIES*sizeof(struct mvpd_toc_entry))
```

krystian-hebel commented on src/soc/ibm/power9/tor.c:
```diff
@@ -0,0 +1,500 @@
+	ring_size = be16toh(ring->size);
+
+	memcpy(ring_section + put_info.ring_offset, &ring_offset, ring_offset);
+	memcpy(ring_section + *ring_section_size, ring, ring_size);
```
These look suspicious. I assume `put_info.ring_offset` and `ring_section_size` are in bytes, and not in multiples of `tor_hdr` sizes. Also, first `memcpy` copies a lot of data from stack and even above stack.

```suggestion
	memcpy((void*)ring_section + put_info.ring_offset, &ring_offset, sizeof(ring_offset));
	memcpy((void*)ring_section + *ring_section_size, ring, ring_size);
```

krystian-hebel commented on src/soc/ibm/power9/tor.h:
```diff
@@ -0,0 +1,446 @@
+};
+
+/* PPE types, enum values matches indecies inside rings section */
+enum ppe_type {
```
I think we should get rid of `sys_phase` and leave just this enum, they seem redundant.

krystian-hebel commented on src/soc/ibm/power9/tor.h:
```diff
@@ -0,0 +1,446 @@
+
+/* Describes ring search characteristics for tor_fetch_and_insert_vpd_rings() */
+struct ring_query {
+	const char *kwd_name;		// Keyword name
```
Is the length of name limited? If it is something sane (say, 8-16 characters or less), put an array here directly.
Also, as this structure is internal to code (not saved externally), please order the fields by size to keep padding low.

krystian-hebel commented on src/soc/ibm/power9/rs4.c:
```diff
@@ -0,0 +1,828 @@
+/* SPDX-License-Identifier: Apache-2.0 */
```
Maybe add a comment that this was copied without modifications with a link to source?

SergiiDmytruk committed:
soc/power9/tor.c: tor_fetch_and_insert_vpd_rings()

This changes previous version to tor_fetch_and_insert_vpd_ring() and
provides function that adds all the necessary rings (ignoring EX which
are unused according to analysis)

Change-Id: I2005c4f0a24615f67a4041c35ef529a6bb60a645
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel commented on src/soc/ibm/power9/tor.c:
```diff
@@ -0,0 +1,769 @@
+
+#define UNDEFINED_PPE_TYPE (uint8_t)0xff
+
+#define NUM_OF_CORES   (uint8_t)24
+#define NUM_OF_QUADS   (uint8_t)6
+#define CORES_PER_QUAD (NUM_OF_CORES/NUM_OF_QUADS)
```
These will have to be moved to some higher level header, but I think it can be left for later.
